registry/registry.go

```diff
@@ -36,6 +40,71 @@ func newTLSConfig(hostname string, isSecure bool) (*tls.Config, error) {
 		}
 	}
 
+	// PKCS11 support: You MUST compile with `make dynbinary`
+	// for this to work. You can then put your locally built
+	// binary into service with:
+	//   sudo systemctl stop docker.service && sudo cp bundles/dynbinary-daemon/dockerd /usr/sbin/dockerd 
+	//
+	// We put the configuration into environment variables as a quick
+	// and dirty way to demonstrate that this works.
+	//
+	// The docker.service file on Debian installs includes:
+	//   EnvironmentFile=-/etc/default/docker
+	//
+	// So you can put something like the following into /etc/default/docker:
+	//
+	//   PKCS11_LABEL=my.cert.name
+	//   PKCS11_PIN=112233
+	//   PKCS11_MODULE=/usr/lib/x86_64-linux-gnu/opensc-pkcs11.so
+	//
+	// You can use only one of PKCS11_SERIAL, PKCS11_SLOT or
+	// PKCS11_LABEL.
+	//
+	// Then `systemctl restart docker.service` to make it take effect.	
+	pkcs11_module := os.Getenv("PKCS11_MODULE")
```

Comment:
I'm also not sure if this is the right location to do this; this code is all on the daemon side, 


- to make this all work, we should also look at containerd code, as well as BuildKit, as all of them may be interacting with registries
- should authentication here happen on the daemon side, or something to be passed from the client?

Reply:
I agree that you would normally expect this to come from the client (same for the private key for client authentication, which is also read here and arguably should not be). Doing that would be a major change, which I don't have the appetite to take on, and I don't think the project would accept anyway. (There has already once before been an effort to *remove* the dependency on miekg/pkcs11, in fact.)

While developing this, I did find myself single stepping far into containerd, but finally I found that I could get `docker pull` and `docker build` working with only these changes, and that's what I needed to get unblocked on my own problem.

I decided to share this now to have this exact conversation.